doc/AUTOMATED_TEST.md

```diff
@@ -0,0 +1,40 @@
+
+## What Code Tests Are Performed
+
+- Unit tests: this includes every set of codes that serves a unique functionality. Unit test eveolves as new codes are added.
+- [BMI](https://bmi.readthedocs.io/en/stable/) (Basical Model Interface) based formulations tests including codes in C, C++, Fortran, and Python.
+- Running **ngen** executable on example hydrofabric with various realistic modules/models, initial condition, and forcing data.
+
+## Local Testing
+
+We strongly recommend you run all the tests on your local computer before you push up the new codes to the branch in your fork. To be able to run the tests, when building the executables (for `build` with `cmake` see [here](https://github.com/stcui007/ngen/blob/ngen_automated_test/doc/BUILDS_AND_CMAKE.md)), you need to set the test option to `ON` using `cmake` option:
+
+    -DNGEN_WITH_TESTS:BOOL=ON
+
+After `build` completes, assuming your build directory is `cmake_build`, you can check that the `ngen` executable is in `cmake_build` directory, and all `unit test` executables are in `cmake_build/test` directory. To run a unit test, for example, run the following command while in the top project directory:
+
+    ./cmake_build/test/test_unit
+
+There are many other unit test executables you need to run for a complete test.
+To run a ngen test job, for example, using the data set in the `data` directory, and run the following command:
+
+    ./cmake_build/ngen data/catchment_data.geojson '' data/nexus_data.geojson '' data/example_bmi_multi_realization_config.json
+
+To run a multi-processors job with MPI, please see a complete description in [here](https://github.com/stcui007/ngen/blob/ngen_automated_test/doc/DISTRIBUTED_PROCESSING.md)
+
+## What to Do When a Test Fail
+
+- Before getting into debugging, first thing, we recommend that you perform all the necessary tests listed above on you local computer. If all tests are successful, then push up your codes to the intended branch in your fork.
+
+- Sometimes, some tests may fail even they have passed tests on local computer. If that happens, you have to look into details why they failed. To do that, click on the word `Details` in blue on the right for a particular test. This will open a window with detailed information for that particular test, the error information are usually near the bottom. You can scroll up and down the side bar for more information. you can also search by key workds in `Search logs` menu entry at the upper right corner.
```

Comment:
```suggestion
- Sometimes, tests may fail even when they have passed locally. If that happens, you have to look into the details of why they failed. To do that, click on the word `Details` in blue on the right for a particular test. This will open a window with detailed information for that particular test, the error information are usually near the bottom. You can scroll up and down the side bar for more information. You can also search by key words in `Search logs` menu entry at the upper right corner.
```